### PR TITLE
[backend] kubernetes: fix secret size limitation

### PR DESCRIPTION
**Pull request by k0da**
By now kubernetes backend could hold up to defaultETCDSize gzipped data
(which is 1-1.5Mb). This doesn't scale for larger states.

This commit implements spliting data across multiple secrets bound by
the same Secret labels. This practically removes etcd value size
limitation and allows backend to scale across multiple secrets.

This also takes care of cases when state needs to be shrinked. In such
case code will cleanup unneeded secrets.

Signed-off-by: Dinar Valeev <dinar.valeev@absa.africa>

### CODE REVIEW

**Review by kuritka — APPROVED** [review]

**Comment by Skaronator** [comment]
This is basically a similar fix as https://github.com/hashicorp/terraform/pull/28838, but for the kubernetes state file.

**Comment by crw** [comment]
Thanks for this submission! I will raise this with the internal maintainers of the kubernetes backend. Although I cannot commit to having this PR reviewed at this time, we acknowledge your contribution and appreciate it! Thanks again for the submission!

**Comment by Bernix01** [comment]
Just making noise to this PR, is this a ready to go?

File: internal/backend/remote-state/kubernetes/client.go
Comment (alexsomesan) [inline internal/backend/remote-state/kubernetes/client.go:150]: AFAIK the Create call translates into a POST HTTP method. How does this behave when the Secret for the chunk already exists?

File: internal/backend/remote-state/kubernetes/client.go
Comment (alexsomesan) [inline internal/backend/remote-state/kubernetes/client.go:187]: Is this not the same as `chunks = append(chunks, buf)`? 
Unsure what the purpose of the slice indexing is...

**Review by alexsomesan — COMMENTED** [review]
It was a bit difficult to comprehend certain parts of this change so I left some spot questions that might hopefully clarify it.

Also, a  more general question: is there potential for conflict if more than one terraform workspace is stored in the same K8s namespace?

File: internal/backend/remote-state/kubernetes/client.go
Comment (jrhouston) [reply internal/backend/remote-state/kubernetes/client.go]: It was used at the end of the function to build the name for the delete call for secrets that were no longer used. I've moved this to where it is actually used.

File: internal/backend/remote-state/kubernetes/client.go
Comment (jrhouston) [reply internal/backend/remote-state/kubernetes/client.go:146]: The logic above this checks if the secret already exists or not and only calls create if it does not already exist. We could however reformulate this as a single patch call instead. This is logic that was already present for the single secret backend, it has just been shuffled around in this PR.

File: internal/backend/remote-state/kubernetes/client.go
Comment (jrhouston) [reply internal/backend/remote-state/kubernetes/client.go]: Yep this was unneeded – removed.

**Comment by jrhouston** [comment]
> Also, a more general question: is there potential for conflict if more than one terraform workspace is stored in the same K8s namespace?

The purpose of the `secret_suffix` here is so you can store multiple terraform states in the same namespace and the secrets will get a different name. You would cause problems if you deliberately used the same prefix across more than one terraform config.

**Review by alexsomesan — APPROVED** [review]
Thanks for fixes @jrhouston! 
Looks good!

**Comment by Bobonium** [comment]
Was there a specific reason why this was not merged and released after @alexsomesan approved it in february? This feature would still be very much appreciated.

**Comment by crw** [comment]
Looks like it got missed. I will add it to the triage queue. Thanks!

**Comment by kmoe** [comment]
Thanks for your patience, everyone!

**Comment by sftim** [comment]
When the data are gzip encoded, what I'd expect to find is that each Secret holds a valid gzip stream. Concatenating these also produces a gzip stream that you can uncompress and deserialize.

Is that how it works?

**Comment by sftim** [comment]
:thought_balloon: I wonder whether and how we provide atomic updates when there are multiple Secrets

**Pull request merged**
